src/main/java/build/buildfarm/worker/CASFileCache.java

```diff
@@ -272,21 +279,33 @@ private void fetchDirectory(
     }
   }
 
-  public synchronized Path putDirectory(
+  public Path putDirectory(
       Digest digest,
       Map<Digest, Directory> directoriesIndex) throws IOException, InterruptedException {
     Path path = getDirectoryPath(digest);
 
-    DirectoryEntry e = directoryStorage.get(digest);
-    if (e != null) {
-      incrementReferences(e.inputs);
-      return path;
+    synchronized (this) {
+      DirectoryEntry e = directoryStorage.get(digest);
+      if (e != null) {
+        incrementReferences(e.inputs);
+        return path;
+      }
     }
 
     ImmutableList.Builder<Path> inputsBuilder = new ImmutableList.Builder<>();
-    fetchDirectory(digest, path, digest, directoriesIndex, inputsBuilder);
+    try {
+      fetchDirectory(digest, path, digest, directoriesIndex, inputsBuilder);
+    } catch (IOException e) {
+      ImmutableList<Path> inputs = inputsBuilder.build();
+      synchronized (this) {
+        purgeDirectoryFromInputs(digest, inputs);
+        decrementReferencesSynchronized(inputs, ImmutableList.<Digest>of());
```

Comment:
Wait, so who calls directoryStorage.remove(digest)? It should be called, right?

Reply:
we haven't yet added this digest to the directoryStorage at this point, this is all that is required to effect the rollback. The only call to `directoryStorage.remove` is in `expireDirectory`

Reply:
/facepalm
yep, you're right, sorry!